File: aoc/2017/aoc201705.go
package main

import (
	"fmt"
	"os"
)

func readInput() []int {
	file, err := os.Open("input/05.txt")
	if err != nil {
		panic(err.Error())
	}
	defer file.Close()
	result := []int{}
	for {
		var i int
		if n, _ := fmt.Fscanf(file, "%d\n", &i); n == 0 {
			return result
		}
		result = append(result, i)
	}
}

func jumps(offsets []int) int {
	nsteps := 0
	i := 0
	for {
		if i < 0 || i >= len(offsets) {
			return nsteps
		}
		nexti := i + offsets[i]
		nsteps++
		offsets[i]++
		i = nexti
	}
}

func jumps2(offsets []int) int {
	nsteps := 0
	i := 0
	for {
		if i < 0 || i >= len(offsets) {
			return nsteps
		}
		nexti := i + offsets[i]
		nsteps++
		if offsets[i] >= 3 {
			offsets[i]--
		} else {
			offsets[i]++
		}
		i = nexti
	}
}

func main() {
	if n := jumps([]int{0, 3, 0, 1, -3}); n != 5 {
		panic(fmt.Sprintf("%d != 5", n))
	}
	if n := jumps2([]int{0, 3, 0, 1, -3}); n != 10 {
		panic(fmt.Sprintf("%d != 10", n))
	}
	println(jumps(readInput()))
	println(jumps2(readInput()))
}

File: aoc/2017/aoc201705.hs
parse :: String -> ([Int],[Int])
parse s = (map read $ words s,[])

jumps :: Int -> ([Int],[Int]) -> Int
jumps nsteps ([],_) = nsteps
jumps nsteps (offset:forward,backward)
  | offset >= 0 = jumps (nsteps+1) (newForward,reverse skipForward ++ backward)
  | -offset > length backward = nsteps+1
  | otherwise = jumps (nsteps+1) (reverse skipBackward ++ offset+1:forward,newBackward)
  where (skipForward,newForward) = splitAt offset (offset+1:forward)
        (skipBackward,newBackward) = splitAt (-offset) backward

test :: ()
test
  | jumps 0 ([0,3,0,1,-3],[]) /= 5 = error "a"
  | otherwise = ()

part1 :: IO Int
part1 = fmap (jumps 0 . parse) $ readFile "input/05.txt"

-- fails with stack overflow
jumps2 :: Int -> ([Int],[Int]) -> Int
jumps2 nsteps ([],_) = nsteps
jumps2 nsteps (offset:forward,backward)
  | offset >= 0 = jumps2 (nsteps+1) (newForward,reverse skipForward ++ backward)
  | -offset > length backward = nsteps+1
  | otherwise = jumps2 (nsteps+1) (reverse skipBackward ++ newOffset:forward,newBackward)
  where (skipForward,newForward) = splitAt offset (newOffset:forward)
        (skipBackward,newBackward) = splitAt (-offset) backward
        newOffset | offset >= 3 = offset - 1 | otherwise = offset + 1

test2 :: ()
test2
  | jumps2 0 ([0,3,0,1,-3],[]) /= 10 = error "a"
  | otherwise = ()

part2 :: IO Int
part2 = fmap (jumps2 0 . parse) $ readFile "input/05.txt"
